chalk-engine/src/logic.rs

```diff
@@ -199,175 +142,737 @@ impl<C: Context> Forest<C> {
+            next_subgoal.answer_index.increment();
+            let next_strand = Strand {
+                infer: strand.infer.clone(),
+                ex_clause: strand.ex_clause.clone(),
+                selected_subgoal: Some(next_subgoal),
+                last_pursued_time: strand.last_pursued_time.clone(),
+            };
+            let table = self.stack[*depth].table;
+            let canonical_next_strand = Self::canonicalize_strand(next_strand);
+            self.tables[table].push_strand(canonical_next_strand);
+        }
+
+        // We got an answer for this goal. Now it matters if this
+        // was a positive or negative subgoal.
+
+        // We are done with this subgoal selection, so we can take it
+        let selected_subgoal = strand.selected_subgoal.take().unwrap();
+        // We've already taken care of the case where we might be able
+        // to pursue this subgoal later (in a different strand). For
+        // this strand, we have an answer to this subgoal
+        match strand
+            .ex_clause
+            .subgoals
+            .remove(selected_subgoal.subgoal_index)
+        {
+            Literal::Positive(subgoal) => {
+                // OK, let's follow *this* answer and see where it leads.
+
+                let SelectedSubgoal {
+                    subgoal_index: _,
+                    subgoal_table,
+                    answer_index,
+                    ref universe_map,
+                } = selected_subgoal;
+                let table_goal = &C::map_goal_from_canonical(
+                    &universe_map,
+                    &C::canonical(&self.tables[subgoal_table].table_goal),
+                );
+                let answer_subst = &C::map_subst_from_canonical(
+                    &universe_map,
+                    &self.answer(subgoal_table, answer_index).subst,
+                );
+                match strand.infer.apply_answer_subst(
+                    &mut strand.ex_clause,
+                    &subgoal,
+                    table_goal,
+                    answer_subst,
+                ) {
+                    Ok(()) => {
+                        let Strand {
+                            infer,
+                            ex_clause,
+                            selected_subgoal: _,
+                            last_pursued_time: _,
+                        } = strand;
+
+                        // If the answer had delayed literals, we have to
+                        // ensure that `ex_clause` is also delayed. This is
+                        // the SLG FACTOR operation, though NFTD just makes it
+                        // part of computing the SLG resolvent.
+                        if self.answer(subgoal_table, answer_index).ambiguous {
+                            ex_clause.ambiguous = true;
+                        }
+
+                        // Increment time counter because we received a new answer.
+                        ex_clause.current_time.increment();
```

Comment:
I agree that having both can be a bit confusing. I'll change the name to `answer_time`. However, I'm not sure if there's a *better* way to do this, since one is bound to answers for a `Strand`, and the other really tracks the "clock" of pursuing an answer on the `Forest`. I'll try to also clarify the docs around these too.